File: tutorials/csharp/loading_balance/lb_server/dbpushpeer.cs
using System;
using SocketProAdapter;
using SocketProAdapter.ServerSide;
using System.Data;

public class DBPushPeer : CAdoClientPeer
{
    public DBPushPeer()
    {
        OnAdonetLoaded += OnAdoNet;
    }

    [RequestAttr(CAdoSerializationHelper.idDmlTriggerMessage, true)]
    private void OnDmlTriggerMessage(int triggerEvent, string fullDbObjectName, object param)
    {
        Console.WriteLine("FullDbTableName = {0}, event = {1}, param = {2}", fullDbObjectName, triggerEvent, param);
    }

    [RequestAttr(CAdoSerializationHelper.idRecordsetName, true)]
    private void OnQueryMessage(string recordsetName)
    {
        Console.WriteLine("Recordset = " + recordsetName);
    }

    [RequestAttr(CAdoSerializationHelper.idDbEventTriggerMessage, true)]
    private void OnDbTriggerMessage(int triggerEvent, string instance, string eventData)
    {
        Console.WriteLine("DB Logon event = " + instance + ", eventData = " + eventData);
    }

    private void OnAdoNet(CAdoClientPeer peer, ushort reqId)
    {
        switch (reqId)
        {
            case CAdoSerializationHelper.idDataReaderRecordsArrive:
                if (AdoSerializer.CurrentDataTable.Rows.Count > 10 * 1024)
                {
                    AdoSerializer.CurrentDataTable.Clear();
                }
                break;
            default:
                break;
        }

        if (reqId == CAdoSerializationHelper.idEndDataReader)
        {
            if (AdoSerializer.CurrentDataTable.Rows.Count > 100)
            {
                Console.WriteLine("Table rowset size = " + AdoSerializer.CurrentDataTable.Rows.Count);
            }
            else
            {
                foreach (DataRow dr in AdoSerializer.CurrentDataTable.Rows)
                {
                    int n = 0;
                    foreach (object obj in dr.ItemArray)
                    {
                        if (n > 0)
                            Console.Write(",\t");
                        Console.Write(obj.ToString());
                        ++n;
                    }
                    Console.WriteLine();
                }
            }
        }
    }
}

File: tutorials/vbnet/loading_balance/lb_server/dbpushpeer.vb
Imports SocketProAdapter
Imports SocketProAdapter.ServerSide

Public Class DBPushPeer
	Inherits CAdoClientPeer

	Public Sub New()
		AddHandler OnAdonetLoaded, AddressOf OnAdoNet
	End Sub

	<RequestAttr(CAdoSerializationHelper.idDmlTriggerMessage, True)>
	Private Sub OnDmlTriggerMessage(ByVal triggerEvent As Integer, ByVal fullDbObjectName As String, ByVal param As Object)
		Console.WriteLine("FullDbTableName = {0}, event = {1}, param = {2}", fullDbObjectName, triggerEvent, param)
	End Sub

	<RequestAttr(CAdoSerializationHelper.idRecordsetName, True)>
	Private Sub OnQueryMessage(ByVal recordsetName As String)
		Console.WriteLine("Recordset = " & recordsetName)
	End Sub

	<RequestAttr(CAdoSerializationHelper.idDbEventTriggerMessage, True)>
	Private Sub OnDbTriggerMessage(ByVal triggerEvent As Integer, ByVal instance As String, ByVal eventData As String)
		Console.WriteLine("DB Logon event = " & instance & ", eventData = " & eventData)
	End Sub

	Private Sub OnAdoNet(ByVal peer As CAdoClientPeer, ByVal reqId As UShort)
		Select Case reqId
			Case CAdoSerializationHelper.idDataReaderRecordsArrive
				If AdoSerializer.CurrentDataTable.Rows.Count > 10 * 1024 Then
					AdoSerializer.CurrentDataTable.Clear()
				End If
			Case Else
		End Select

		If reqId = CAdoSerializationHelper.idEndDataReader Then
			If AdoSerializer.CurrentDataTable.Rows.Count > 100 Then
				Console.WriteLine("Table rowset size = " & AdoSerializer.CurrentDataTable.Rows.Count)
			Else
				For Each dr As DataRow In AdoSerializer.CurrentDataTable.Rows
					Dim n As Integer = 0
					For Each obj As Object In dr.ItemArray
						If n > 0 Then
							Console.Write("," & ControlChars.Tab)
						End If
						Console.Write(obj.ToString())
						n += 1
					Next obj
					Console.WriteLine()
				Next dr
			End If
		End If
	End Sub
End Class